gpflow/kernels/stationaries.py

```diff
@@ -46,35 +46,75 @@ def ard(self) -> bool:
         """
         return self.lengthscales.shape.ndims > 0
 
-    def scaled_squared_euclid_dist(self, X, X2=None):
-        """
-        Returns ||(X - X2ᵀ) / ℓ||² i.e. squared L2-norm.
-        """
-        X_scaled = X / self.lengthscales
-        X2_scaled = X2 / self.lengthscales if X2 is not None else X2
-        return square_distance(X_scaled, X2_scaled)
+    def scale(self, X):
+        X_scaled = X / self.lengthscales if X is not None else X
+        return X_scaled
+
+    def K_diag(self, X):
+        return tf.fill(tf.shape(X)[:-1], tf.squeeze(self.variance))
+
+
+class IsotropicStationary(Stationary):
+    """
+    Base class for isotropic stationary kernels, i.e. kernels that only
+    depend on
+
+        d = ‖x - x'‖
+
+    Derived classes should implement one of:
+
+        K_r2(self, r2): Returns the kernel evaluated on r² (r2), which is the
+        squared scaled Euclidean distance Should operate element-wise on r².
+
+        K_r(self, r): Returns the kernel evaluated on r, which is the scaled
+        Euclidean distance Should operate element-wise on r².
+    """
 
     def K(self, X, X2=None):
         r2 = self.scaled_squared_euclid_dist(X, X2)
         return self.K_r2(r2)
 
-    def K_diag(self, X):
-        return tf.fill(tf.shape(X)[:-1], tf.squeeze(self.variance))
-
     def K_r2(self, r2):
-        """
-        Returns the kernel evaluated on r² (`r2`), which is the squared scaled Euclidean distance
-        Should operate element-wise on r²
-        """
         if hasattr(self, "K_r"):
             # Clipping around the (single) float precision which is ~1e-45.
             r = tf.sqrt(tf.maximum(r2, 1e-40))
-            return self.K_r(r)  # pylint: disable=no-member
+            return self.K_r(r)
+        raise NotImplementedError
+
+    def scaled_squared_euclid_dist(self, X, X2=None):
+        """
+        Returns ||(X - X2ᵀ) / ℓ||² i.e. squared L2-norm.
+        """
+        return square_distance(self.scale(X), self.scale(X2))
+
+
+class AnisotropicStationary(Stationary):
+    """
+    Base class for anisotropic stationary kernels, i.e. kernels that only
+    depend on
 
+        d = x - x'
+
+    Derived classes should implement K_d(self, d): Returns the kernel evaluated
+    on d, which is pairwise difference matrix, scaled by the lengthscale
+    parameter ℓ (i.e. Σ_i [(X - X2ᵀ) / ℓ]_i). The last axis corresponds to
+    input dimension.
+    """
+
+    def K(self, X, X2=None):
+        return self.K_d(self.scaled_distance(X, X2))
+
+    def scaled_distance(self, X, X2=None):
```

Comment:
as in the utilities module: "distance" would let me expect a scalar - how about calling this `scaled_difference_matrix`?

what would its return shape be?